File: src/main/java/org/evernet/service/ConfigService.java
package org.evernet.service;

import lombok.RequiredArgsConstructor;
import org.evernet.bean.Vertex;
import org.evernet.exception.NotFoundException;
import org.evernet.model.Config;
import org.evernet.repository.ConfigRepository;
import org.evernet.util.Random;
import org.springframework.stereotype.Service;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ConfigService {

    private final ConfigRepository configRepository;

    private static final String JWT_SIGNING_KEY = "jwtSigningKey";
    private static final String VERTEX_ENDPOINT = "vertexEndpoint";
    private static final String VERTEX_DISPLAY_NAME = "vertexDisplayName";
    private static final String VERTEX_DESCRIPTION = "vertexDescription";
    private static final String FEDERATION_PROTOCOL = "federationProtocol";

    public void init(Vertex vertex) {
        insert(VERTEX_ENDPOINT, vertex.getEndpoint());
        insert(VERTEX_DISPLAY_NAME, vertex.getDisplayName());
        insert(VERTEX_DESCRIPTION, vertex.getDescription());
        insert(JWT_SIGNING_KEY, Random.generateRandomString(128));
        insert(FEDERATION_PROTOCOL, "http");
    }

    public void insert(String key, String value) {
        configRepository.insertIfNotExists(UUID.randomUUID().toString(), key, value);
    }

    public void set(String key, String value) {
        configRepository.upsert(UUID.randomUUID().toString(), key, value);
    }

    public Config get(String key) {
        Config config = configRepository.findByKey(key);

        if (config == null) {
            throw new NotFoundException(String.format("Config %s not found", key));
        }

        return config;
    }

    public String getJwtSigningKey() {
        return get(JWT_SIGNING_KEY).getValue();
    }

    public void setJwtSigningKey() {
        set(JWT_SIGNING_KEY, Random.generateRandomString(128));
    }

    public String getVertexEndpoint() {
        return get(VERTEX_ENDPOINT).getValue();
    }

    public void setVertexEndpoint(String endpoint) {
        set(VERTEX_ENDPOINT, endpoint);
    }

    public String getVertexDisplayName() {
        return get(VERTEX_DISPLAY_NAME).getValue();
    }

    public void setVertexDisplayName(String displayName) {
        set(VERTEX_DISPLAY_NAME, displayName);
    }

    public String getVertexDescription() {
        return get(VERTEX_DESCRIPTION).getValue();
    }

    public void setVertexDescription(String description) {
        set(VERTEX_DESCRIPTION, description);
    }

    public String getFederationProtocol() {
        return get(FEDERATION_PROTOCOL).getValue();
    }

    public void setFederationProtocol(String protocol) {
        set(FEDERATION_PROTOCOL, protocol);
    }
}

File: service/config_service.py
import datetime

from model.config import Config
from utils import secret


class ConfigService:
    def __init__(self):
        pass

    @staticmethod
    def init():
        ConfigService.upsert(key='jwt_signing_key', value=secret.generate(64))

    @staticmethod
    def upsert(key: str, value: str):
        Config.insert(
            key=key,
            value=value,
            created_at=datetime.datetime.now(tz=datetime.timezone.utc),
            updated_at=datetime.datetime.now(tz=datetime.timezone.utc)
        ).on_conflict(
            "IGNORE"
        ).execute()

    @staticmethod
    def set(key: str, value: str):
        Config.insert(
            key=key,
            value=value,
            created_at=datetime.datetime.now(tz=datetime.timezone.utc),
            updated_at=datetime.datetime.now(tz=datetime.timezone.utc)
        ).on_conflict(
            conflict_target=[Config.key],
            preserve=[Config.created_at],
            update={
                Config.value: value,
                Config.updated_at: datetime.datetime.now(tz=datetime.timezone.utc)
            }
        ).execute()

    @staticmethod
    def get_or_default(key: str, default: str) -> str:
        val = Config.select(Config.value).where(Config.key == key).get_or_none()

        if val is None:
            return default

        return val.value

    @staticmethod
    def get_jwt_signing_key():
        return ConfigService.get_or_default('jwt_signing_key', 'secret')

    @staticmethod
    def set_vertex_endpoint(vertex_endpoint: str):
        ConfigService.set("vertex_endpoint", vertex_endpoint)

    @staticmethod
    def get_vertex_endpoint():
        return ConfigService.get_or_default('vertex_endpoint', "localhost:3000")

    @staticmethod
    def set_vertex_display_name(vertex_display_name: str):
        ConfigService.set('vertex_display_name', vertex_display_name)

    @staticmethod
    def get_vertex_display_name():
        return ConfigService.get_or_default('vertex_display_name', "Vertex")

    @staticmethod
    def set_vertex_description(vertex_description):
        ConfigService.set('vertex_description', vertex_description)

    @staticmethod
    def get_vertex_description():
        return ConfigService.get_or_default('vertex_description', "Vertex")

    @staticmethod
    def get_federation_protocol():
        return ConfigService.get_or_default('federation_protocol', "http")
